File: WordParser.py
import re
import os
from DBAdapter import *

# This program reads text scripts from the audio_script table and produces the audio_words table.
# It outputs the text as token types and tokens.  The following are the token types:
# W Word: this is whole words, including hypenated words
# P Punctuation: this is single punctuation characters
# S Whitespace: this is exact whitespace, and can be multiple characters.
# V Verse number: {n} or {n}_{n} this is the exact character string as shown.
# N Number (was considered, but not done)
# While languages might differ, this program needs a consistent way to decide what is a word.
# It assumes that whitespace is always a word delimiter, but a single character of punctuation
# is not a word delimiter.  But, it assumes that multiple characters of punctuation within a word
# is word delimiter.


class WordParser:

	def __init__(self, db):
		self.db = db
		self.word_seq = 0
		self.lastScriptId = None


	def parse(self):
		BEGIN = 1
		SPACE = 2
		WORD = 3 
		WORDPUNCT = 4
		VERSENUM = 5
		INVERSENUM = 6
		ENDVERSENUM = 7
		NEXTVERSENUM = 8
		label = ["", "BEGIN", "SPACE", "WORD", "WORDPUNCT", "VERSENUM", "INVERSENUM", "ENDVERSENUM", 
			"NEXTVERSENUM"] 
		for (script_id, usfm_style, verse_num, script_text) in self.db.selectScripts():
			print(script_text)
			term = None
			punct = None
			verseeStr = None
			state = BEGIN
			for token in re.split(r"([\W])", script_text):
				if len(token) > 0:
					print(label[state], "now token: ", token)
					if state == BEGIN:
						if token.isspace():
							term = token
							state = SPACE
						elif token.isalnum(): 
							term = token
							state = WORD   
						elif token == '{':
							term = token
							state = VERSENUM
						else: # token.ispunct()
							self.addWord(script_id, verse_num, 'P', token)
							term = None # redundant
							state = BEGIN
					elif state == SPACE:
						if token.isspace():
							term += token
						elif token.isalnum():
							self.addWord(script_id, verse_num, 'S', term)
							term = token
							state = WORD
						elif token == '{':
							self.addWord(script_id, verse_num, 'S', term)
							term = token
							state = VERSENUM
						else: # token.ispunct()
							self.addWord(script_id, verse_num, 'S', term)
							self.addWord(script_id, verse_num, 'P', token)
							term = None
							state = BEGIN
					elif state == WORD:
						if token.isspace():
							self.addWord(script_id, verse_num, 'W', term)
							term = token 
							state = SPACE
						elif token.isalnum():
							self.logError("space or punct", token)
						elif token == '{':
							self.addWord(script_id, verse_num, 'W', term)
							term = token
							state = VERSENUM
						else: # token.ispunct()
							punct = token 
							state = WORDPUNCT
					elif state == WORDPUNCT:
						if token.isspace():
							self.addWord(script_id, verse_num, 'W', term)
							self.addWord(script_id, verse_num, 'P', punct)
							punct = None
							term = token 
							state = SPACE
						elif token.isalnum():
							term += punct + token 
							punct = None 
							state = WORD
						else: # token.ispunct()
							self.addWord(script_id, verse_num, 'W', term)
							self.addWord(script_id, verse_num, 'P', punct)
							self.addWord(script_id, verse_num, 'P', token)
							term = None
							state = BEGIN
					elif state == VERSENUM:
						if token.isdigit():
							term += token 
							verseStr = token
							state = INVERSENUM
						else:
							self.logError("number", token)
					elif state == INVERSENUM:
						if token.isdigit():
							term += token
							verseStr += token
						elif token == '}':
							term += token
							verse_num = int(verseStr)
							state = ENDVERSENUM
						else:
							self.logError("number or }", token)
					elif state == ENDVERSENUM:
						if token == '_':
							term += token 
							state = NEXTVERSENUM
						elif token.isspace():
							self.addWord(script_id, verse_num, 'V', term)
							term = token
							state = SPACE
						elif token.isalnum():
							self.addWord(script_id, verse_num, 'V', term)
							term = token 
							state = WORD 
						else: # token.ispunct()
							print("output verse", term)
							self.addWord(script_id, verse_num, 'V', term)
							self.addWord(script_id, verse_num, 'P', token)
							term = None 
							state = BEGIN
					elif state == NEXTVERSENUM:
						if token == '{':
							term += token
							state = INVERSENUM
						else:
							self.logError("{", token)

			if term != None and len(term) > 0:
				if term.isspace():
					self.addWord(script_id, verse_num, 'S', term)
				elif term.isalnum():
					self.addWord(script_id, verse_num, 'W', term)
		self.db.insertWords()


	def addWord(self, script_id, verse_num, ttype, text):
		if self.lastScriptId != script_id:
			self.lastScriptId = script_id
			self.word_seq = 0
		self.word_seq += 1
		if ttype == None or text == None:# or verse_num == None:
			print(script_id, verse_num, ttype, text)
			sys.exit(0)
		print("seq: ", self.word_seq, " versee: ", verse_num, " type: ", ttype, " text: ", text)
		self.db.addWord(script_id, self.word_seq, verse_num, ttype, text)


	def logError(self, expected, actual):
		print("Expected: ", expected, ", but found: ", actual)
		sys.exit(1)


	def format(self, filename):
		with open(filename, "w") as file:
			sql = """SELECT s.script_id, s.book_id, s.chapter_num, w.verse_num, w.word_seq, w.ttype, w.word 
				FROM audio_scripts s JOIN audio_words w ON s.script_id = w.script_id ORDER BY w.word_id"""
			for (script_id, book_id, chapter_num, verse_num, word_seq, ttype, word) in db.sqlite.select(sql):
				file.write(word)


if __name__ == "__main__":
	db = DBAdapter("ZAK_MWRIGHT.db")
	db.deleteWords()
	word = WordParser(db)
	word.parse()
	word.format("ZAK_MWRIGHT_WORDS.txt")
	with open("ZAK_MWRIGHT_SCRIPT.txt", "w") as file:
		sql = "SELECT book_id, chapter_num, in_verse_num, script_text FROM audio_scripts ORDER BY script_id"
		for (book_id, chapter_num, in_verse_num, script_text) in db.sqlite.select(sql):
			file.write(script_text)

File: read/word_parser.go
package read

import (
	"context"
	"dataset"
	"dataset/db"
	log "dataset/logger"
	"strconv"
	"unicode"
)

/**
This program reads text scripts from the script table and produces the words table.
It outputs the text as term types and terms.  The following are the term types:
W Word: this is whole words, including hypenated words
P Punctuation: this is single punctuation characters
S Whitespace: this is exact whitespace, and can be multiple characters.
V Verse number: {n} or {n}_{n} this is the exact character string as shown.
While languages might differ, this program needs a consistent way to decide what is a word.
It assumes that whitespace is always a word delimiter, but a single character of punctuation
is not a word delimiter.  But, it assumes that multiple characters of punctuation within a word
is a word delimiter.
*/

type WordParser struct {
	ctx          context.Context
	conn         db.DBAdapter
	wordSeq      int
	lastScriptId int
	records      []db.Word
}

func NewWordParser(conn db.DBAdapter) WordParser {
	var w WordParser
	w.ctx = conn.Ctx
	w.conn = conn
	w.wordSeq = 0
	w.lastScriptId = 0
	return w
}

func (w *WordParser) Parse() dataset.Status {
	const (
		begin = iota + 1
		space
		word
		wordPunct
		verseNum
		inVerseNum
		endVerseNum
		nextVerseNum
	)
	var label = []string{"", "BEGIN", "SPACE", "WORD", "WORDPUNCT", "VERSENUM", "INVERSENUM", "ENDVERSENUM",
		"NEXTVERSENUM"}
	var records, status = w.conn.SelectScripts()
	if status.IsErr {
		return status
	}
	for _, rec := range records {
		//fmt.Printf("%s %d:%s  %s\n", rec.BookId, rec.ChapterNum, rec.VerseStr, rec.ScriptText)
		var term = make([]rune, 0, 100) // None
		var punct rune                  // None
		var verseStr []rune             // None
		var state = begin
		for _, tok := range rec.ScriptText {
			//fmt.Printf("%s\t%d\t%c\t%d\n", label[state], pos, tok, tok)
			switch state {
			case begin:
				if unicode.IsSpace(tok) {
					term = append(term, tok)
					state = space
				} else if unicode.IsLetter(tok) || unicode.IsNumber(tok) {
					term = append(term, tok)
					state = word
				} else if tok == '{' {
					term = append(term, tok)
					state = verseNum
				} else { // IsPunct
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{tok})
					term = []rune{} // redundant
					state = begin
				}
			case space:
				if unicode.IsSpace(tok) {
					term = append(term, tok)
				} else if unicode.IsLetter(tok) || unicode.IsNumber(tok) {
					w.addWord(rec.ScriptId, rec.VerseNum, `S`, term)
					term = []rune{tok}
					state = word
				} else if tok == '{' {
					w.addWord(rec.ScriptId, rec.VerseNum, `S`, term)
					term = []rune{tok}
					state = verseNum
				} else { // IsPunct
					w.addWord(rec.ScriptId, rec.VerseNum, `S`, term)
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{tok})
					term = []rune{}
					state = begin
				}
			case word:
				if unicode.IsSpace(tok) {
					w.addWord(rec.ScriptId, rec.VerseNum, `W`, term)
					term = []rune{tok}
					state = space
				} else if unicode.IsLetter(tok) || unicode.IsNumber(tok) {
					term = append(term, tok)
				} else if tok == '{' {
					w.addWord(rec.ScriptId, rec.VerseNum, `W`, term)
					term = []rune{tok}
					state = verseNum
				} else { // IsPunct
					punct = tok
					state = wordPunct
				}
			case wordPunct:
				if unicode.IsSpace(tok) {
					w.addWord(rec.ScriptId, rec.VerseNum, `W`, term)
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{punct})
					punct = -1
					term = []rune{tok}
					state = space
				} else if unicode.IsLetter(tok) || unicode.IsNumber(tok) {
					term = append(term, punct)
					term = append(term, tok)
					punct = -1
					state = word
				} else { // IsPunct
					w.addWord(rec.ScriptId, rec.VerseNum, `W`, term)
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{punct})
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{tok})
					term = []rune{}
					state = begin
				}
			case verseNum:
				if unicode.IsDigit(tok) {
					term = append(term, tok)
					verseStr = []rune{tok}
					state = inVerseNum
				} else {
					w.logError("number", tok)
				}
			case inVerseNum:
				if unicode.IsDigit(tok) {
					term = append(term, tok)
					verseStr = append(verseStr, tok)
				} else if tok == '}' {
					term = append(term, tok)
					rec.VerseNum, _ = strconv.Atoi(string(verseStr))
					verseStr = []rune{}
					state = endVerseNum
				} else {
					w.logError("number or }", tok)
				}
			case endVerseNum:
				if tok == '_' {
					term = append(term, tok)
					state = nextVerseNum
				} else if unicode.IsSpace(tok) {
					w.addWord(rec.ScriptId, rec.VerseNum, `V`, term)
					term = []rune{tok}
					state = space
				} else if unicode.IsLetter(tok) || unicode.IsNumber(tok) {
					w.addWord(rec.ScriptId, rec.VerseNum, `V`, term)
					term = []rune{tok}
					state = word
				} else { // IsPunct
					w.addWord(rec.ScriptId, rec.VerseNum, `V`, term)
					w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{tok})
					term = []rune{}
					state = begin
				}
			case nextVerseNum:
				if tok == '{' {
					term = append(term, tok)
					state = inVerseNum
				} else {
					w.logError("{", tok)
				}
			default:
				return log.ErrorNoErr(w.ctx, 500, "unknown state", label[state])
			}
		}
		if len(term) > 0 && term[0] != -1 {
			var first = term[0]
			if unicode.IsSpace(first) {
				w.addWord(rec.ScriptId, rec.VerseNum, `S`, term)
			} else if unicode.IsLetter(first) || unicode.IsNumber(first) {
				w.addWord(rec.ScriptId, rec.VerseNum, `W`, term)
			} else { // IsPunct
				w.addWord(rec.ScriptId, rec.VerseNum, `P`, term)
			}
		}
		if state == wordPunct && punct != -1 {
			w.addWord(rec.ScriptId, rec.VerseNum, `P`, []rune{punct})
		}
	}
	w.conn.DeleteWords()
	status = w.conn.InsertWords(w.records)
	w.records = []db.Word{}
	return status
}

func (w *WordParser) addWord(scriptId int, verseNum int, ttype string, text []rune) dataset.Status {
	var status dataset.Status
	if w.lastScriptId != scriptId {
		w.lastScriptId = scriptId
		w.wordSeq = 0
	}
	w.wordSeq += 1
	if ttype == `` || len(text) == 0 { // or rec.VerseNum == None:
		return log.ErrorNoErr(w.ctx, 500, 0, `Aparant bug trying to addWord`)
	}
	//fmt.Println("seq: ", w.wordSeq, " verse: ", verseNum, " type: ", ttype, " word: ", string(text))
	var rec db.Word
	rec.ScriptId = scriptId
	rec.WordSeq = w.wordSeq
	rec.VerseNum = verseNum
	rec.TType = ttype
	rec.Word = string(text)
	w.records = append(w.records, rec)
	return status
}

func (w *WordParser) logError(expected string, actual rune) dataset.Status {
	return log.ErrorNoErr(w.ctx, 500, "Expected: ", expected, ", but found: ", string(actual))
}
